src/jaxsim/api/references.py

```diff
@@ -441,3 +442,111 @@ def convert_using_link_frame(
         return replace(
             forces=self.input.physics_model.f_ext.at[link_idxs, :].set(W_f0_L + W_f_L)
         )
+
+    def apply_frame_forces(
+        self,
+        forces: jtp.MatrixLike,
+        model: js.model.JaxSimModel,
+        data: js.data.JaxSimModelData,
+        frame_names: tuple[str, ...] | str | None = None,
+        additive: bool = False,
+    ) -> Self:
+        """
+        Apply the frame forces.
+
+        Args:
+            forces: The frame 6D forces in the active representation.
+            model:
+                The model to consider, only needed if a frame serialization different
+                from the implicit one is used.
+            data:
+                The data of the considered model, only needed if the velocity
+                representation is not inertial-fixed.
+            frame_names: The names of the frames corresponding to the forces.
+            additive:
+                Whether to add the forces to the existing ones instead of replacing them.
+
+        Returns:
+            A new `JaxSimModelReferences` object with the given frame forces.
+
+        Note:
+            The frame forces must be expressed in the active representation.
+            Then, we always convert and store forces in inertial-fixed representation.
+        """
+
+        f_F = jnp.atleast_2d(forces).astype(float)
+
+        # If we have the model, we can extract the frame names if not provided.
+        frame_names = frame_names if frame_names is not None else model.frame_names()
+
+        # Make sure that the frame names are a tuple if they are provided by the user.
+        frame_names = (frame_names,) if isinstance(frame_names, str) else frame_names
+
+        if len(frame_names) != f_F.shape[0]:
+            msg = "The number of frame names ({}) must match the number of forces ({})"
+            raise ValueError(msg.format(len(frame_names), f_F.shape[0]))
+
+        # Extract the frame indices.
+        frame_idxs = js.frame.names_to_idxs(frame_names=frame_names, model=model)
+        parent_link_idxs = js.frame.idx_of_parent_link(
+            model=model, frame_idxs=frame_idxs
+        )
+
+        if not_tracing(forces) and not data.valid(model=model):
+            raise ValueError("The provided data is not valid for the model")
+
+        W_H_Fi = jax.vmap(
+            lambda frame_idx: js.frame.transform(
+                model=model, data=data, frame_index=frame_idx
+            )
+        )(frame_idxs)
+
+        # Helper function to convert a single 6D force to the inertial representation
+        # considering as body the frame (i.e. L_f_F and LW_f_F).
+        def to_inertial(
+            f_F: jtp.MatrixLike, W_H_F: jtp.MatrixLike, frame_idx: jtp.ArrayLike
```

Comment:
It seems to me that the frame index is unused, can we remove it?

Reply:
Totally yes! Having [F841](https://docs.astral.sh/ruff/rules/unused-variable/) disabled [here](https://github.com/ami-iit/jaxsim/blob/4189a2c4e794ec3e191e660cc9f2285522fe9ce1/pyproject.toml#L159) makes this type of error invisible. Addressed in https://github.com/ami-iit/jaxsim/pull/209/commits/cf74cfffcd101e1d08f802c03e14ec6f078e9129

Reply:
I added https://github.com/ami-iit/jaxsim/pull/209/commits/f30d3cf84bade8a4e35671c085609a7736f53cfb to prevent this error in the future. I needed to add a lot of `noqa`, so it kinda makes the code base a bit less clean. Let me know if you want me to drop it

Reply:
Having the `noqa` is not a problem on my side, I prefer keeping the extra assigned variable for readability, it saves (at least to me) mental bandwidth when I review the code -even if it's mine xD-.

Reply:
I totally agree with that!